dso-l2/src/test/java/com/tc/objectserver/handler/RetirementManagerTest.java

```diff
@@ -317,6 +317,36 @@ public void testRetirementWithMultiMessageDependency() throws Exception {
     // deferRequest2 completion will retire both invokeRequest and deferRequest2
     Assert.assertThat(toRetire, IsIterableContainingInOrder.contains(invokeRequest, deferRequest1));
   }
+  
+  @Test
+  public void testChainedRetirement() throws Exception {
+    Retiree invokeRequest1 = makeResponse();
+    EntityMessage invokeMessage1 = mock(EntityMessage.class);    
+    
+    Retiree invokeRequest2 = makeResponse();
+    EntityMessage invokeMessage2 = mock(EntityMessage.class);
+
+    Retiree invokeRequest3 = makeResponse();
+    EntityMessage invokeMessage3 = mock(EntityMessage.class);
+    
+    registerWithMessage(invokeRequest1, invokeMessage1, 1);
+    
+    this.retirementManager.deferRetirement(invokeMessage1, invokeMessage2);
+    
+    this.retirementManager.deferRetirement(invokeMessage2, invokeMessage3);
+    
+    registerWithMessage(invokeRequest2, invokeMessage2, 1);
+
+    registerWithMessage(invokeRequest3, invokeMessage3, 1);
+    
+    List<Retiree> retireList = this.retirementManager.retireForCompletion(invokeMessage1);
+    
+    retireList = this.retirementManager.retireForCompletion(invokeMessage2);
```

Comment:
Also it would be to good if you can add some test with messages on different concurrency keys.